specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1086 @@
+    Nonce             uint64  
+    // BottomUpNonce of bottomup messages for msgMeta received from che    kpoints.
+    // This nonce is used to mark with a nonce the metadata about cross-net
+    // messages received in checkpoints. This is used to order the
+    // bottom-up cross-messages received through checkpoints.
+    BottomUpNonce     uint64
+    // Queue of bottom-up cross-net messages to be applied.
+    BottomUpMsgsMeta  cid.Cid // AMT[schema.CrossMsgs]
+    // AppliedNonces keep track of the next nonce of the message to be applied.
+    // This prevents potential replay attacks.
+    AppliedBottomUpNonce uint64
+    AppliedTopDownNonce  uint64
+    // Registry with all active atomic executions being orchestrated
+    // by the current subnet.
+    AtomicExecRegistry cid.Cid // HAMT[cid]AtomicExec
+}
+```
+
+> Struct with the information the SCA keeps for each child subnet
+```go
+// Subnet struct kept by the SCA with the information of
+// all of its children.
+type Subnet struct {
+    // ID of the Subnet
+    ID          address.SubnetID 
+    // Parent ID
+    ParentID    address.SubnetID
+    //  Collateral staked for this subnet.
+    Collateral       abi.TokenAmount
+    // List of cross top-down messages committed for the subnet..
+    TopDownMsgs cid.Cid // AMT[ltypes.Messages] 
+    // Latest nonce of cross message submitted to subnet.
+    Nonce      uint64
+    // Amount of native tokens injected in the subnet and
+    // that can be used freely in the subnet.
+    CircSupply abi.TokenAmount 
+    // Status of the checkpoint (`Active`, `Inactive`, etc.)
+    Status     Status
+    // Latest checkpoint committed for the subnet.
+    // Kept for verification purposes.
+    PrevCheckpoint schema.Checkpoint
+}
+```
+
+## Consensus Interface
+Each subnet can run its own implementation of a consensus algorithm. In order for different consensus implementations to operate semlessly with the Filecoin stack, we decouple the core methods of the current consensus in its own interace. In order for a consensus implementation to be usuable in a subnet it needs to implement this interface.
+> Consensus Interface
+```go
+type Consensus interface {
+    // Performs a complete check of a proposed block to either
+    // accept it or reject it.
+    ValidateBlock(ctx context.Context, b *types.FullBlock) (err error)
+    // Light check performed when a block is received through 
+    // the pubsub channel.
+    ValidateBlockPubsub(ctx context.Context, self bool, msg *pubsub.Message) (pubsub.ValidationResult, string)
+
+    // Used by mining processes to assemble and propose a new
+    // signed block.
+    CreateBlock(ctx context.Context, w api.Wallet, bt *api.BlockTemplate) (*types.FullBlock, error)
+    // Return the type of consensus. 
+    Type() hierarchical.ConsensusType
+}
+```
+An example implementation of a BFT-like consensus protocol for HC can be found [here](https://github.com/filecoin-project/mir)
```

Comment:
This might be a bit confusing, since the Mir actually does not implement that interface, as this is very far from the interface a BFT-style consensus would have. Instead, there is code in Eudico itself implementing that interface and using the Mir library under the hood.

Reply:
Yes. We'll need to add a section in the future to describe more accurately how the implementation of new consensus algorithm works (describing the role of the consensus interface, the ordering interface, maybe the mempool, etc.).